tfhe/src/core_crypto/algorithms/lwe_compact_public_key_generation.rs

```diff
@@ -0,0 +1,149 @@
+    generator.fill_slice_with_random_mask(mask.as_mut());
+
+    slice_semi_reverse_negacyclic_convolution(
+        body.as_mut(),
+        mask.as_ref(),
+        lwe_secret_key.as_ref(),
+    );
+
+    generator
+        .unsigned_torus_slice_wrapping_add_random_noise_assign(body.as_mut(), noise_parameters);
+}
+
+/// Allocate a new [`LWE compact public key`](`LweCompactPublicKey`) and fill it with an actual
+/// public key constructed from a private [`LWE secret key`](`LweSecretKey`).
+///
+/// See [`encrypt_lwe_ciphertext_with_compact_public_key`] for usage.
+pub fn allocate_and_generate_new_lwe_compact_public_key<Scalar, InputKeyCont, Gen>(
+    lwe_secret_key: &LweSecretKey<InputKeyCont>,
+    noise_parameters: impl DispersionParameter,
+    ciphertext_modulus: CiphertextModulus<Scalar>,
+    generator: &mut EncryptionRandomGenerator<Gen>,
+) -> LweCompactPublicKeyOwned<Scalar>
+where
+    Scalar: UnsignedTorus,
+    InputKeyCont: Container<Element = Scalar>,
+    Gen: ByteRandomGenerator,
+{
+    let mut pk = LweCompactPublicKeyOwned::new(
+        Scalar::ZERO,
+        lwe_secret_key.lwe_dimension(),
+        ciphertext_modulus,
+    );
+
+    generate_lwe_compact_public_key(lwe_secret_key, &mut pk, noise_parameters, generator);
+
+    pk
+}
+
+/// Fill a [`seeded LWE compact public key`](`LweCompactPublicKey`) with an actual public key
+/// constructed from a private [`LWE secret key`](`LweSecretKey`).
+pub fn generate_seeded_lwe_compact_public_key<Scalar, InputKeyCont, OutputKeyCont, NoiseSeeder>(
+    lwe_secret_key: &LweSecretKey<InputKeyCont>,
+    output: &mut SeededLweCompactPublicKey<OutputKeyCont>,
+    noise_parameters: impl DispersionParameter,
+    noise_seeder: &mut NoiseSeeder,
+) where
+    Scalar: UnsignedTorus,
+    InputKeyCont: Container<Element = Scalar>,
+    OutputKeyCont: ContainerMut<Element = Scalar>,
+    // Maybe Sized allows to pass Box<dyn Seeder>.
+    NoiseSeeder: Seeder + ?Sized,
+{
+    assert!(
+        output.ciphertext_modulus().is_native_modulus(),
+        "This operation only supports native moduli"
+    );
+
+    assert!(
+        lwe_secret_key.lwe_dimension() == output.lwe_dimension(),
+        "Mismatched LweDimension between input LweSecretKey {:?} \
+    and ouptut LweCompactPublicKey {:?}",
+        lwe_secret_key.lwe_dimension(),
+        output.lwe_dimension()
+    );
+
+    let mut generator = EncryptionRandomGenerator::<ActivatedRandomGenerator>::new(
+        output.compression_seed().seed,
+        noise_seeder,
+    );
+
+    let mut tmp_mask = vec![Scalar::ZERO; output.lwe_dimension().0];
+    generator.fill_slice_with_random_mask(tmp_mask.as_mut());
+
+    let mut body = output.get_mut_body();
+
+    slice_semi_reverse_negacyclic_convolution(
+        body.as_mut(),
+        tmp_mask.as_ref(),
+        lwe_secret_key.as_ref(),
+    );
+
+    generator
+        .unsigned_torus_slice_wrapping_add_random_noise_assign(body.as_mut(), noise_parameters);
```

Comment:
Would it make sense here to call `allocate_and_generate_new_seeded_lwe_compact_public_key` and copy the body once done ?

Reply:
I miss copied the name, I meant `allocate_and_generate_new_lwe_compact_public_key`

````rust
 let mut generator = EncryptionRandomGenerator::<ActivatedRandomGenerator>::new(
        output.compression_seed().seed,
        noise_seeder,
    );

let pk = allocate_and_generate_new_lwe_compact_public_key(
    lwe_secre_key,
    noise_parameter,
    generator,
    output.ciphertext_modulu
);

*output.get_mut_body = pk.get_body();
```

Reply:
but that would allocate a full key 🤔

Reply:
yeah and the key would not be seeded

basically you have two functions, one which fills an existing entity, there is a specifc version for a seeded and normal entity and then the allocate and generate new which creates the entity and calls the generation impl as needed

Reply:
Why would they not be seeded ? We are giving the generator that uses the given seed

Reply:
allocate_and_generate_new_lwe_compact_public_key returns the expanded entity so not seeded in the sense it's not compressed